src/libcsg/modules/io/pdbreader.cc

```diff
@@ -57,128 +61,383 @@ void PDBReader::Close()
+                        top.CreateResidue("DUMMY"); // create dummy residue, hopefully it will never show
+                        cout << "Warning: residue numbers not continous, create DUMMY residue with nr " << top.ResidueCount() << endl;
+                    }
+                    top.CreateResidue(resName);
+                }
+                // This is not correct, but still better than no type at all!
+                BeadType *type = top.GetOrCreateBeadType(atName);
+
+                // Determine if the charge has been provided in the .pdb file or if we will
+                // be assuming it is 0
+                double ch=0;
+                if(charge!=""){
+                    ch = boost::lexical_cast<double>(charge);
+                }
+                // CreateBead takes 6 parameters in the following order
+                // 1 - symmetry of the bead (1-indicates sphere, 3-indicates ellipsoidal)
+                // 2 - name of the bead     (string)
+                // 3 - bead type            (BeadType *)
+                // 4 - residue number       (int)
+                // 5 - mass                 (double)
+                // 6 - charge               (double)
+                //
+                // res -1 as internal number starts with 0
+                b = top.CreateBead(1, atName, type, resnr-1,this->getMass(atName),ch);
+            }else{
+                b = top.getBead(bead_count-1);
+            }
             // convert to nm from A
             b->setPos(vec(
-                    boost::lexical_cast<double>(x)/10.0,
-                    boost::lexical_cast<double>(y)/10.0,
-                    boost::lexical_cast<double>(z)/10.0
-                ));
+                        boost::lexical_cast<double>(x)/10.0,
+                        boost::lexical_cast<double>(y)/10.0,
+                        boost::lexical_cast<double>(z)/10.0
+                        ));
+
+            bead_vec.push_back(b);
 
-	}
+        }
 
         if (( line == "ENDMDL" ) || ( line == "END" ) || ( _fl.eof())){
-          break;
-	}
+            break;
+        }
     }
 
-    if(!_topology && (i>0) && i != top.BeadCount())
-      throw std::runtime_error("number of beads in topology and trajectory differ");
-   
-    if (_topology)
-      cout << "WARNING: topology created from .pdb file, charges and masses are wrong!\n";
+    if(!_topology && (bead_count>0) && bead_count != top.BeadCount())
+        throw std::runtime_error("number of beads in topology and trajectory differ");
+
+
+    ////////////////////////////////////////////////////////////////////////////////
+    // Sort data and determine atom structure, connect with top (molecules, bonds)
+    ////////////////////////////////////////////////////////////////////////////////
+
+    // Extra processing is done if the file is a topology file, in which case the atoms
+    // must be sorted into molecules and the bond interactions recorded
+    if(_topology)
```

Comment:
I suppose that this part of the code is working correctly. But I think you could avoid all those comments by using other data structures. In fact what you need is a list of atoms in every molecule (e.g. again some ```unordered_map<int, list<int> >```) and for every atom id the corresponding molecule id. (e.g. ```map<int, int>```).

Reply:
Hmm ok, you are exposing me to objects I am not as familiar with... this is very good. I will see what I can do.